experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,338 @@
+          val port = appConfig.getEnv(YARNAPPMASTER_PORT).toInt
+          val target = host + ":" + port
+          val addr = NetUtils.createSocketAddr(target)
+          rmClientActor ! RegisterAMMessage(addr.getHostName, port, trackingURL)
+        case Success(false) =>
+          LOG.error("Failed to start AMRMClientAsync")
+        case Failure(ex) =>
+          LOG.error("Failed to start AMRMClientAsync", ex)
+        case _ =>
+          LOG.error("Unknown status for AMRMClientAysncStartup")
+      }
+
+    case containerStarted: ContainerStarted =>
+      LOG.info(s"Started container : ${containerStarted.containerId}")
+      if (needMoreMasterContainersState) {
+        masterContainersStarted += 1
+        LOG.info(s"Currently master containers started : $masterContainersStarted/${appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt}")
+        requestWorkerContainersIfNeeded()
+      } else {
+        workerContainersStarted += 1
+        LOG.info(s"Currently worker containers started : $workerContainersStarted/${appConfig.getEnv(WORKER_CONTAINERS).toInt}")
+        servicesActor match {
+          case None if workerContainersStarted == workerContainersRequested =>
+            val masters = masterContainers.map(pair => {
+              val (_, (host, port)) = pair
+              host + ":" + port
+            }).toArray
+            servicesActor = Some(context.actorOf(Props(classOf[ServicesLauncherActor], masters, host, servicesPort)))
+          case _ =>
+        }
+      }
+
+    case containerRequest: ContainerRequestMessage =>
+      LOG.info("AM: Received ContainerRequestMessage")
+      rmClientActor ! containerRequest
+
+    case amResponse: RegisterAppMasterResponse =>
+      LOG.info("Received RegisterAppMasterResponse")
+      requestMasterContainers(amResponse.response)
+
+    case containers: LaunchContainers =>
+      LOG.info("Received LaunchContainers")
+      if (needMoreMasterContainersState) {
+        LOG.info(s"Launching more masters : ${containers.containers.size}")
+        setMasterAddrIfNeeded(containers.containers)
+        launchMasterContainers(containers.containers)
+      } else if (needMoreWorkerContainersState) {
+        LOG.info(s"Launching more workers : ${containers.containers.size}")
+        workerContainersRequested += containers.containers.size
+        launchWorkerContainers(containers.containers)
+      } else {
+        LOG.info("No more needed")
+      }
+
+    case done: RMHandlerDone =>
+      LOG.info("Got RMHandlerDone")
+      cleanUp(done)
+
+    case unknown =>
+      LOG.info(s"Unknown message ${unknown.getClass.getName}")
+
+  }
+
+  private def setMasterAddrIfNeeded(containers: List[Container]) {
+    masterAddr match {
+      case None =>
+        masterAddr = Some(HostPort(containers.head.getNodeId.getHost, appConfig.getEnv(GEARPUMPMASTER_PORT).toInt))
+      case _ =>
+    }
+
+  }
+
+  private def needMoreMasterContainersState: Boolean = {
```

Comment:
what needMoreMasterContainersState mean?
It means need more master contaner?
Or need more master container state?